### PR TITLE
PCIE: add PTM root support and API to enable PTM requester

### PR DESCRIPTION
**Pull request by tbursztyka**
I don't think of any use atm about having a PCIe port/bridge driver so instead the instance is just a PTM root (if relevant). It's not affecting any part of the code, besides DTS.

In the future, if we support a more "runtime friendly" way, we will be able to have a pcie_ptm_root_enable() api instead (which would relevantly enable the right root device dynamically) and no entry in DTS at all.

### CODE REVIEW

**Review by superna9999 — COMMENTED** [review]
While I agree we will only support PCIe, these defines are taken from the PCI Code
and ID Assignment Specification Revision 1.11 document.

I agree we can rename the extended, which are only PCIe, but the non-extended capabilites are PCI numbers, and the PCI Express® Base Specification Revision 5.0 clearly name then PCI capabilities.

**Comment by tbursztyka** [comment]
@superna9999 ok I'll keep PCI_ for the legacy capabilities

**Review by superna9999 — APPROVED** [review]

**Review by dcpleung — APPROVED** [review]

File: drivers/pcie/host/Kconfig [hunk 1]
```diff
@@ -44,6 +44,12 @@ config PCIE_MSI_X
 
 endif # PCIE_MSI
 
+config PCIE_PTM
+	bool "Enable support for PCI(e) Precision Time Management (PTM)"
```
Comment (finikorg) [inline drivers/pcie/host/Kconfig:48]: Is this `Precision Time Measurement` ?

File: drivers/pcie/host/ptm.c [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+#include <errno.h>
+
+#include <kernel.h>
+
+#include <soc.h>
+#include <device.h>
+#include <init.h>
+
+#include <drivers/pcie/pcie.h>
+#include "ptm.h"
```
Comment (finikorg) [inline drivers/pcie/host/ptm.c:20]: BTW, do we need to have this header?

File: drivers/pcie/host/ptm.c [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+
+static int pcie_ptm_root_setup(const struct device *dev, uint32_t base)
+{
+	const struct pcie_ptm_root_config *config = dev->config;
+	union ptm_cap_reg cap;
+	union ptm_ctrl_reg ctrl;
+
+	cap.raw = pcie_conf_read(config->bdf, base + PTM_CAP_REG_OFFSET);
+	if ((cap.root == 0) || ((cap.root == 1) && cap.responder == 0)) {
```
Comment (finikorg) [inline drivers/pcie/host/ptm.c:29]: `if ((cap.root == 0) || ((cap.root == 1) && (cap.responder == 0))) {` ?

Comment (tbursztyka) [reply drivers/pcie/host/ptm.c:29]: indeed

File: drivers/pcie/host/Kconfig (hunk 1 above)
Comment (tbursztyka) [reply drivers/pcie/host/Kconfig:48]: No it's Precision Time Management , at least from PCIe 4.0 r1 specs (section 7.9.16)

File: drivers/pcie/host/ptm.c (hunk 2 above)
Comment (tbursztyka) [reply drivers/pcie/host/ptm.c:20]: Yes

File: drivers/pcie/host/Kconfig (hunk 1 above)
Comment (finikorg) [reply drivers/pcie/host/Kconfig:48]: I think in most of the places they use  name `Measurements` in that document

Comment (superna9999) [reply drivers/pcie/host/Kconfig:48]: In `PCI Express® Base Specification Revision 5.0 Version 1.0`, the capability name is indeed `Precision Time Management Extended Capability` for controlling `PTM` Hierarchy which means `Precision Time Measurement`.

Comment (finikorg) [reply drivers/pcie/host/Kconfig:48]: In that document the seems to use both names for that capability, so it should be OK to use `Management`.

**Review by jhedberg — APPROVED** [review]

File: drivers/pcie/host/ptm.c [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+
+#define PCIE_PTM_ROOT_INIT(index)					\
+	static const struct pcie_ptm_root_config ptm_config_##index = {	\
+		.bdf = DT_INST_REG_ADDR(index),				\
+	};								\
+	DEVICE_DT_INST_DEFINE(index, &pcie_ptm_root_init, NULL, NULL,	\
+			      &ptm_config_##index, PRE_KERNEL_1,	\
+			      CONFIG_KERNEL_INIT_PRIORITY_DEVICE, NULL);
+
+DT_INST_FOREACH_STATUS_OKAY(PCIE_PTM_ROOT_INIT)
```
Comment (mbolivar-nordic) [inline drivers/pcie/host/ptm.c:66]: Where is DT_DRV_COMPAT being set for this to work?

**Pull request merged**